`CollapseConsecutiveAssertThatStatements` does not work on `Integers` (possibly only on `List`s?)

<!--
Thank you for reporting an issue with OpenRewrite!
We appreciate you taking the time to help us improve.
Please fill out the template below to help us understand and reproduce the issue.
Feel free to delete any sections that don't apply to your issue.
-->

## What version of OpenRewrite are you using?
<!--
Whenever possible please try to replicate your issue with the latest versions of OpenRewrite.
The latest major and minor versions of OpenRewrite projects are usually listed here:
https://docs.openrewrite.org/reference/latest-versions-of-every-openrewrite-module
For patch releases check the GitHub Releases page for the respective project.

We release every few weeks, so it's possible that your issue has already been fixed.

If you want to try the most recent changes that haven't been fully released yet, you can check out our snapshot releases
https://docs.openrewrite.org/reference/snapshot-instructions
-->
I am using the latest as of 2024-09-20

## How are you running OpenRewrite?
Gradle by editing build file, confirmed by tests on this repository:
- https://github.com/openrewrite/rewrite-testing-frameworks/pull/604 

## What is the smallest, simplest way to reproduce the problem?
<!--
Sometimes the logs indicate a recipe stumbled over a particular pattern of code.
If you can share a code snippet that reproduces the issue, that will help us fix it faster.
We also accept [pull requests that merely replicate an issue](https://github.com/openrewrite/.github/blob/main/CONTRIBUTING.md#contributing-fixes), as a step up to a full fix.

A code snippet can be something simple like this, or similar for other languages:
-->
- see here: https://github.com/openrewrite/rewrite-testing-frameworks/pull/604/

in short:
```java
                      Integer i = 1+1;
                      assertThat(i).isNotNull();
                      assertThat(i).isEqualTo(2);
```

## What did you expect to see?
<!-- A code snippet, or a description of the behavior you expected helps us write a test to ensure the issue is fixed. -->
```java
                      Integer i = 1+1;
                      assertThat(i).isNotNull().isEqualTo(2);
```

## What did you see instead?
no change
## What is the full stack trace of any errors you encountered?
<!-- When errors occur, please include the output of `--stacktrace` for Gradle or `--debug` for Maven. -->
per https://github.com/openrewrite/rewrite-testing-frameworks/actions/runs/10971127404/job/30465846209?pr=604#step:5:1484

```
CollapseConsecutiveAssertThatStatementsTest > collapseAssertThatsOnInteger() FAILED
    java.lang.AssertionError: Recipe was expected to make a change but made no changes.
        at org.openrewrite.test.LargeSourceSetCheckingExpectedCycles.afterCycle(LargeSourceSetCheckingExpectedCycles.java:119)
        at org.openrewrite.RecipeScheduler.runRecipeCycles(RecipeScheduler.java:98)
        at org.openrewrite.RecipeScheduler.scheduleRun(RecipeScheduler.java:41)
        at org.openrewrite.Recipe.run(Recipe.java:376)
        at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:371)
        at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:130)
        at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:125)
        at org.openrewrite.java.testing.assertj.CollapseConsecutiveAssertThatStatementsTest.collapseAssertThatsOnInteger(CollapseConsecutiveAssertThatStatementsTest.java:197)
```
## Context

The existing test are all on `List`s, so maybe the recipe currently only works on those and other types need to be enabled in general.

## Are you interested in [contributing a fix to OpenRewrite](https://github.com/openrewrite/.github/blob/main/CONTRIBUTING.md#contributing-fixes)?
<!-- Indicate if this is something you would like to work on, and how we can best support you in doing so. -->
No

--- a/src/main/java/org/openrewrite/java/testing/assertj/CollapseConsecutiveAssertThatStatements.java
+++ b/src/main/java/org/openrewrite/java/testing/assertj/CollapseConsecutiveAssertThatStatements.java
@@ -127,7 +127,9 @@ public class CollapseConsecutiveAssertThatStatements extends Recipe {
                                     }
                                 }
 
-                                return TypeUtils.isOfType(assertThatFq.getType(), assertionFq.getType());
+                                if (assertThatFq != null && assertionFq != null) {
+                                    return TypeUtils.isOfType(assertThatFq.getType(), assertionFq.getType());
+                                }
                             }
                             return false;
                         }

--- a/src/main/java/org/openrewrite/java/testing/assertj/CollapseConsecutiveAssertThatStatements.java
+++ b/src/main/java/org/openrewrite/java/testing/assertj/CollapseConsecutiveAssertThatStatements.java
@@ -105,7 +105,31 @@ public class CollapseConsecutiveAssertThatStatements extends Recipe {
                         if (!(assertThatArgument instanceof MethodCall)) {
                             JavaType assertThatType = assertThat.getType();
                             JavaType assertionType = assertion.getType();
-                            return TypeUtils.isOfType(assertThatType, assertionType);
+
+                            // Check if both types are the same or if the assertion type is assignable to assertThat type
+                            // This handles cases where assertion methods return the same type (for fluent chaining)
+                            // but avoids collapsing when methods like extracting() change the assertion type
+                            if (assertThatType != null && assertionType != null) {
+                                // First check for exact type match
+                                if (TypeUtils.isOfType(assertThatType, assertionType)) {
+                                    return true;
+                                }
+
+                                // For generic types like AbstractIntegerAssert<?>, we need to check the raw types
+                                JavaType.Parameterized assertThatFq = TypeUtils.asParameterized(assertThatType);
+                                JavaType.Parameterized assertionFq = TypeUtils.asParameterized(assertionType);
+
+                                // If assertionType is a generic wildcard, try to get its bound
+                                if (assertionFq == null && assertionType instanceof JavaType.GenericTypeVariable) {
+                                    JavaType.GenericTypeVariable genericType = (JavaType.GenericTypeVariable) assertionType;
+                                    if (!genericType.getBounds().isEmpty()) {
+                                        assertionFq = TypeUtils.asParameterized(genericType.getBounds().get(0));
+                                    }
+                                }
+
+                                return TypeUtils.isOfType(assertThatFq.getType(), assertionFq.getType());
+                            }
+                            return false;
                         }
                     }
                 }
@@ -117,7 +141,7 @@ public class CollapseConsecutiveAssertThatStatements extends Recipe {
                 Space originalPrefix = consecutiveAssertThatStatement.get(0).getPrefix();
                 String continuationIndent = originalPrefix.getIndent().contains("\t") ? "\t\t" : "        ";
                 Space indentedNewline = Space.format(originalPrefix.getLastWhitespace().replaceAll("^\\s+\n", "\n") +
-                                                     continuationIndent);
+                        continuationIndent);
                 J.MethodInvocation collapsed = null;
                 for (Statement st : consecutiveAssertThatStatement) {
                     J.MethodInvocation assertion = (J.MethodInvocation) st;

--- a/src/test/java/org/openrewrite/java/testing/assertj/CollapseConsecutiveAssertThatStatementsTest.java
+++ b/src/test/java/org/openrewrite/java/testing/assertj/CollapseConsecutiveAssertThatStatementsTest.java
@@ -15,10 +15,10 @@
  */
 package org.openrewrite.java.testing.assertj;
 
-import org.junit.jupiter.api.Disabled;
 import org.junit.jupiter.api.Test;
 import org.openrewrite.DocumentExample;
 import org.openrewrite.InMemoryExecutionContext;
+import org.openrewrite.Issue;
 import org.openrewrite.java.JavaParser;
 import org.openrewrite.test.RecipeSpec;
 import org.openrewrite.test.RewriteTest;
@@ -54,9 +54,6 @@ class CollapseConsecutiveAssertThatStatementsTest implements RewriteTest {
                       assertThat(listA).hasSize(3);
                       assertThat(listA).containsExactly("a", "b", "c");
                   }
-                  private int[] notification() {
-                      return new int[]{1, 2, 3};
-                  }
               }
               """,
             """
@@ -71,9 +68,6 @@ class CollapseConsecutiveAssertThatStatementsTest implements RewriteTest {
                               .isNotNull()
                               .hasSize(3)
                               .containsExactly("a", "b", "c");
-                  }
-                  private int[] notification() {
-                      return new int[]{1, 2, 3};
                   }
               }
               """
@@ -104,10 +98,6 @@ class CollapseConsecutiveAssertThatStatementsTest implements RewriteTest {
                       assertThat(listB).isNotNull();
                       assertThat(listB).hasSize(3);
                   }
-
-                  private int[] notification() {
-                      return new int[]{1, 2, 3};
-                  }
               }
               """,
             """
@@ -129,10 +119,6 @@ class CollapseConsecutiveAssertThatStatementsTest implements RewriteTest {
                       assertThat(listB)
                               .isNotNull()
                               .hasSize(3);
-                  }
-
-                  private int[] notification() {
-                      return new int[]{1, 2, 3};
                   }
               }
               """
@@ -160,10 +146,6 @@ class CollapseConsecutiveAssertThatStatementsTest implements RewriteTest {
                       assertThat(listB).isNotNull();
                       assertThat(listB).hasSize(3);
                   }
-
-                  private int[] notification() {
-                      return new int[]{1, 2, 3};
-                  }
               }
               """,
             """
@@ -183,17 +165,13 @@ class CollapseConsecutiveAssertThatStatementsTest implements RewriteTest {
                               .isNotNull()
                               .hasSize(3);
                   }
-
-                  private int[] notification() {
-                      return new int[]{1, 2, 3};
-                  }
               }
               """
           )
         );
     }
 
-    @Disabled("Not yet implemented")
+    @Issue("https://github.com/openrewrite/rewrite-testing-frameworks/issues/605")
     @Test
     void collapseAssertThatsOnInteger() {
         //language=java
@@ -224,6 +202,59 @@ class CollapseConsecutiveAssertThatStatementsTest implements RewriteTest {
         );
     }
 
+    @Issue("https://github.com/openrewrite/rewrite-testing-frameworks/issues/605")
+    @Test
+    void collapseAssertThatsOnOtherPrimitiveTypes() {
+        //language=java
+        rewriteRun(
+          java(
+            """
+              import static org.assertj.core.api.Assertions.assertThat;
+
+              class MyTest {
+                  void test() {
+                      String s = "hello";
+                      assertThat(s).isNotNull();
+                      assertThat(s).isEqualTo("hello");
+                      assertThat(s).hasSize(5);
+
+                      Long l = 100L;
+                      assertThat(l).isNotNull();
+                      assertThat(l).isGreaterThan(50L);
+
+                      Boolean b = true;
+                      assertThat(b).isNotNull();
+                      assertThat(b).isTrue();
+                  }
+              }
+              """,
+            """
+              import static org.assertj.core.api.Assertions.assertThat;
+
+              class MyTest {
+                  void test() {
+                      String s = "hello";
+                      assertThat(s)
+                              .isNotNull()
+                              .isEqualTo("hello")
+                              .hasSize(5);
+
+                      Long l = 100L;
+                      assertThat(l)
+                              .isNotNull()
+                              .isGreaterThan(50L);
+
+                      Boolean b = true;
+                      assertThat(b)
+                              .isNotNull()
+                              .isTrue();
+                  }
+              }
+              """
+          )
+        );
+    }
+
     @Test
     void ignoreIfAssertThatOnDifferentVariables() {
         //language=java
@@ -240,10 +271,6 @@ class CollapseConsecutiveAssertThatStatementsTest implements RewriteTest {
                       List<String> listB = Arrays.asList("a", "b", "c");
                       assertThat(listA).isNotNull();
                       assertThat(listB).containsExactly("a", "b", "c");
-                  }
-
-                  private int[] notification() {
-                      return new int[]{1, 2, 3};
                   }
               }
               """
@@ -294,9 +321,6 @@ class CollapseConsecutiveAssertThatStatementsTest implements RewriteTest {
                           .hasSize(3);
                       assertThat(listA).containsExactly("a", "b", "c");
                   }
-                  private int[] notification() {
-                      return new int[]{1, 2, 3};
-                  }
               }
               """
           )
@@ -319,9 +343,6 @@ class CollapseConsecutiveAssertThatStatementsTest implements RewriteTest {
                       assertThat(listA).isNotNull();
                       int x=3;
                       assertThat(listA).hasSize(x);
-                  }
-                  private int[] notification() {
-                      return new int[]{1, 2, 3};
                   }
               }
               """

